File: python/2161_partition_array_according_to_given_pivot.py
class Solution:
    def pivotArray(self, nums: List[int], pivot: int) -> List[int]:
        pivot_count = 0
        less_than = deque()
        greater_than = deque()
        for num in nums:
            if num == pivot:
                pivot_count += 1
            elif num < pivot:
                less_than.append(num)
            else:  # num > pivot
                greater_than.append(num)
        result = [None] * len(nums)
        i = 0
        while len(less_than) > 0:
            result[i] = less_than.popleft()
            i += 1
        while pivot_count > 0:
            result[i] = pivot
            pivot_count -= 1
            i += 1
        while len(greater_than) > 0:
            result[i] = greater_than.popleft()
            i += 1
        return result

File: rust/src/2161_partition_array_according_to_given_pivot.rs
impl Solution {
    pub fn pivot_array(nums: Vec<i32>, pivot: i32) -> Vec<i32> {
        let mut result = Vec::with_capacity(nums.len());
        let mut less_than = 0;
        let mut pivot_count = 0;
        let mut greater_than = 0;

        for num in &nums {
            if *num == pivot {
                pivot_count += 1;
            } else if *num < pivot {
                less_than += 1;
            } else {
                greater_than += 1;
            }
            // Need to size the vector up, capacity != size.
            result.push(0);
        }

        let pivot_start = less_than;
        let gt_start = less_than + pivot_count;
        let mut li_index = 0;
        let mut pivot_index = pivot_start;
        let mut gt_index = gt_start;

        for num in &nums {
            if *num == pivot {
                result[pivot_index] = *num;
                pivot_index += 1;
            } else if *num < pivot {
                result[li_index] = *num;
                li_index += 1;
            } else {
                result[gt_index] = *num;
                gt_index += 1;
            }
        }

        return result;
    }
}
